### PR TITLE
Importing issues with requirements and templates

### PR DESCRIPTION
**Pull request by ansharlubis**
Addressing issues #3139 and #3138.

The following files make up the matrix example in separate modules:
- template_semigroup.f90
- template_monoid.f90
- template_semiring.f90
- template_unitring.f90
- template_field.f90
- template_matrix.f90
- template_matrix_real.f90
- template_matrix_integer.f90

### CODE REVIEW

File: src/libasr/modfile.cpp [hunk 1]
```diff
@@ -7,6 +7,7 @@
 #include <libasr/serialization.h>
 #include <libasr/bwriter.h>
 
+#include <libasr/pickle.h>
```
Comment (certik) [inline src/libasr/modfile.cpp:10]: This might not be needed here.

**Comment by certik** [comment]
Some of the new reference tests are 14K of lines, which will increase the size of the git repository. Do we need to check all that? Or can we just run it via integration tests.

**Comment by certik** [comment]
Does it make sense to keep the original all-in-one-file test (to ensure that still works), as well as the new split tests?

**Comment by certik** [comment]
Other than that, the changes look good to me.

**Comment by ansharlubis** [comment]
@certik Some of the tests are superfluous, I want to combine them first. For some of the bigger examples, integration tests should be enough.

**Comment by ansharlubis** [comment]
@certik How do I add the separate generics matrix examples into `integration_tests/CMakeLists.txt`?

**Comment by certik** [comment]
Look at the modules_15 test, that contains several files as well. You probably need to add a main program, so that it creates an executable that can be run.

**Comment by ansharlubis** [comment]
I see, the problem was the missing main program.

**Comment by ansharlubis** [comment]
I follow the other example, added the extra files into `integration_tests/CMakeLists.txt` and the main program but they still don't link properly.

**Comment by certik** [comment]
We have to debug it.

File: src/libasr/asr_utils.h [hunk 2]
```diff
@@ -564,6 +564,51 @@ static inline std::string type_to_str(const ASR::ttype_t *t)
     }
 }
 
+static inline std::string type_to_str_with_substitution(const ASR::ttype_t *t,
```
Comment (certik) [inline src/libasr/asr_utils.h:567]: Can you document what exactly this function does and how it differs to the other similar function for type comparison?

**Review by certik — APPROVED** [review]
I think this is good enough to merge. Left a comment to improve docs.

File: src/libasr/asr_utils.h (hunk 2 above)
Comment (ansharlubis) [reply src/libasr/asr_utils.h:567]: Got it! I'll add the documentation in a separate pull request.

**Pull request merged**